### ISSUE TITLE
API Proposal: BitArray.CopyTo(int value) / BitArray.From(int value)

### ISSUE THREAD

**Issue by Symbai** [dotnet/corefx#41005]
At the moment there is no way to create a System.Collections.BitArray from a single integer and no way to return the bits to a single integer. I don't know the exactly reason why the API is missing but considering its possible to achieve with a 1 length integer array it wouldn't be too difficult to add such an API.

```
namespace System.Collections
{
    public sealed class  BitArray : ICollection, IEnumerable, ICloneable
    {
        //Fyi: I would have considered BitArray(int value) instead but it's conflicting with BitArray(int length)
        public static From(int value);
        public void CopyTo(int value);
   }
}
```

Example::
```
static void ChangeBit(ref int value,  int position, bool newBit)
{
    var bits = BitArray.From(value);
    bits[position] = newBit;
    bits.CopyTo(value);
}
static int ReplaceBitRange(int existingInteger, int newInteger, int position, int length)
{
    var oldBits = BitArray.From(existingInteger);
    var newBits = BitArray.From(newInteger);
    var counter = 0;
    for (int i = position; i < position + length; i++)
    {
        oldBits[i] = newBits[counter++];
    }
    int result;
    oldBits.CopyTo(result);
    return result;
}
```
Without these APIs we would have to create arrays when initializing the bit array and when returning, which adds more performance cost in scenarios where it matters. While changing bits directly from and to an integer is better.

If getting approved perhaps this can be considered to be not only for integer but also for byte.

**Comment by Clockwork-Muse** [comment]
`BitArray(int length)` is almost certainly due to `<SomeCollectionType>(int length)`.

>Without these APIs we would have to create arrays when initializing the bit array and when returning, which adds more performance cost in scenarios where it matters. While changing bits directly from and to an integer is better.

This implies you have a significant number of pre-initialized integers (or a similar number of destinations), which strikes me as.... unusual.  If you don't have a pre-initialized value, you don't need the value consumption, of course.

**Comment by AlexejLiebenthal** [comment]
I just thought about your problem. And as `BitArray` is implemented as an [integer array field](https://github.com/dotnet/corefx/blob/76f566a281bbe979e80cbbb3a48ddf522cdcb4e1/src/System.Collections/src/System/Collections/BitArray.cs#L19), you need to create the array anyway.

If I understand you correctly, I think, I struggled sometime ago with a similar problem like you.
If you like to have some O(1) bit manipulation abstraction on `int`, I hacked something for you. Especially the `ReplaceBitRange` method should perform better (O(length) vs. O(1)).
```cs
public struct IntBit
{
    private readonly int value;

    public IntBit(int value) => this.value = value;

    public IntBit ChangeBit(int pos, bool bit)
    {
        if (pos < 0 || pos > 31) throw new ArgumentOutOfRangeException(nameof(pos));

        return new IntBit(value | (bit ? 1 : 0) << pos);
    }

    public IntBit ReplaceBitRange(int bits, int start, int length)
    {
        if (start < 0 || start > 31) throw new ArgumentOutOfRangeException(nameof(start));
        if (length < 1 && start + length <= 32) throw new ArgumentOutOfRangeException(nameof(length));

        var mask = ((1 << length) - 1) << start;
        return new IntBit((bits & mask) | (value & ~mask));
    }

    public static implicit operator int(IntBit intBit) => intBit.value;
}

public static class IntBitManipulationExtensions
{
    public static int ChangeBit(this int value, int pos, bool bit) => new IntBit(value).ChangeBit(pos, bit);
    public static int ReplaceBitRange(this int value, int bits, int start, int length) => new IntBit(value).ReplaceBitRange(bits, start, length);
}
```

**Comment by Symbai** [comment]
@Clockwork-Muse @AlexejLiebenthal Thank you, I didn't know about that. For me I always needed BitArray for single integers, not arrays and it's the only way unless you know how to shift like @AlexejLiebenthal . Would you say it makes sense to have a new API for getting bits only from a single integer? Maybe even in BitConverter instead of BitArray? Or leaving this to external nuget packages?

I'm asking because I think I'm not the only one who would need it for single integers only. On the other hand when it comes to performance and this might be the only reason why not declaring an array matters, shifting is more likely the way to go.

**Comment by Wraith2** [comment]
>  Would you say it makes sense to have a new API for getting bits only from a single integer?

Can't you use [BitVector32](https://docs.microsoft.com/en-us/dotnet/api/system.collections.specialized.bitvector32?view=netcore-3.0)

**Comment by Symbai** [comment]
@Wraith2 Thanks for the tip but unfortunately it generates different bits on signed values.
```
const int testValue = -1873524097;
var one = new BitVector32(testValue);
var two = new BitArray(new [] {testValue });
Debug.Assert(one[7] == two[7]); // <---- Throws
```

**Comment by Wraith2** [comment]
That is unexpected. The bit values for the signed version aren't the same either. 
I wasn't suggesting interoperability though i thought you wanted a single int that's addressible via bits and in that case using BitVector32 throughout should do what you want.

**Comment by AlexejLiebenthal** [comment]
`BitVector32` has a different implemenation of `this[int] {get;}`.

`BitArray` checks whether the indexed bit is set or not.
`BitVector` checks whether all bits are set to get the indexed value.
e.g.: 0b0110 -> 6
BitArray[0]
BitArray[0]

**Comment by Wraith2** [comment]
Oh, so you'd have to use bit consts to get the value you wanted, at which point you might as well be just masking yourself.

**Comment by AlexejLiebenthal** [comment]
Yes, it's more like checking via index, if a mask matches the (inner data) of `BitVector32`- instance.

I find the naming `bit` a bit misleading. Instead should this be renamed in my eyes to `mask`

>         // Summary:
>         //     Gets or sets the state of the bit flag indicated by the specified mask.
>         //
>         // Parameters:
>         //   bit:
>         //     A mask that indicates the bit to get or set.
>         //
>         // Returns:
>         //     true if the specified bit flag is on (1); otherwise, false.
>         public bool this[int bit] { get; set; }

**Comment by stephentoub** [comment]
I don't think we should add such an API; it's too specific, and if the goal is to eliminate an array allocation, it still won't help with the fact that the BitArray itself needs to be allocated and an underlying array it manages needs to be allocated.

If we were going to add any APIs here, it should be a ctor that takes a `ReadOnlySpan<byte>` and a `CopyTo` method that takes a `Span<byte>`.  If you wanted to use that to then construct a `BitArray` from an int without allocating an array, you could create a span around the int and reinterpret cast it to a span of bytes.
